{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.append(\"../src/modeval\")\n",
    "from modulescomparison import ModulesComparison\n",
    "from modulecontainers import Modules "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0,os.path.abspath(\"../src/modeval/\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulatorId</th>\n",
       "      <th>regulatorName</th>\n",
       "      <th>RegulatorGeneName</th>\n",
       "      <th>regulatedId</th>\n",
       "      <th>regulatedName</th>\n",
       "      <th>function</th>\n",
       "      <th>confidenceLevel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>RDBECOLIPDC00358</td>\n",
       "      <td>DsrA</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>RDBECOLIGNC00450</td>\n",
       "      <td>hns</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>RDBECOLIPDC00358</td>\n",
       "      <td>DsrA</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>RDBECOLIGNC00539</td>\n",
       "      <td>lrp</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>RDBECOLIPDC00358</td>\n",
       "      <td>DsrA</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>RDBECOLIGNC00804</td>\n",
       "      <td>rbsA</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>RDBECOLIPDC00358</td>\n",
       "      <td>DsrA</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>RDBECOLIGNC00805</td>\n",
       "      <td>rbsB</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>RDBECOLIPDC00358</td>\n",
       "      <td>DsrA</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>RDBECOLIGNC00806</td>\n",
       "      <td>rbsC</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          regulatorId regulatorName RegulatorGeneName       regulatedId  \\\n",
       "313  RDBECOLIPDC00358          DsrA              dsrA  RDBECOLIGNC00450   \n",
       "315  RDBECOLIPDC00358          DsrA              dsrA  RDBECOLIGNC00539   \n",
       "316  RDBECOLIPDC00358          DsrA              dsrA  RDBECOLIGNC00804   \n",
       "317  RDBECOLIPDC00358          DsrA              dsrA  RDBECOLIGNC00805   \n",
       "318  RDBECOLIPDC00358          DsrA              dsrA  RDBECOLIGNC00806   \n",
       "\n",
       "    regulatedName function confidenceLevel  \n",
       "313           hns        -              S   \n",
       "315           lrp        -              S   \n",
       "316          rbsA        -              S   \n",
       "317          rbsB        -              S   \n",
       "318          rbsC        -              S   "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get the known RegulonDB network with modification described in inspect_input_elements.ipynb\n",
    "known_trn = pd.read_csv(\"../data/Regulatory/QCd_Network.csv\", index_col=0)\n",
    "known_trn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b0001</td>\n",
       "      <td>thrL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>thr operon leader peptide</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>P0AD86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337</td>\n",
       "      <td>2799</td>\n",
       "      <td>+</td>\n",
       "      <td>fused aspartate kinase/homoserine dehydrogenase 1</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>P00561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0003</td>\n",
       "      <td>thrB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2801</td>\n",
       "      <td>3733</td>\n",
       "      <td>+</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>P00547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0004</td>\n",
       "      <td>thrC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3734</td>\n",
       "      <td>5020</td>\n",
       "      <td>+</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>P00934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0005</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5234</td>\n",
       "      <td>5530</td>\n",
       "      <td>+</td>\n",
       "      <td>DUF2502 domain-containing protein YaaX</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>P75616</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag gene_name  old_locus_tag  start   end strand  \\\n",
       "0     b0001      thrL            NaN    190   255      +   \n",
       "1     b0002      thrA            NaN    337  2799      +   \n",
       "2     b0003      thrB            NaN   2801  3733      +   \n",
       "3     b0004      thrC            NaN   3734  5020      +   \n",
       "4     b0005      yaaX            NaN   5234  5530      +   \n",
       "\n",
       "                                        gene_product  \\\n",
       "0                          thr operon leader peptide   \n",
       "1  fused aspartate kinase/homoserine dehydrogenase 1   \n",
       "2                                  homoserine kinase   \n",
       "3                                 threonine synthase   \n",
       "4             DUF2502 domain-containing protein YaaX   \n",
       "\n",
       "                                   COG uniprot  \n",
       "0                    No COG annotation  P0AD86  \n",
       "1  Amino acid transport and metabolism  P00561  \n",
       "2  Nucleotide transport and metabolism  P00547  \n",
       "3  Amino acid transport and metabolism  P00934  \n",
       "4                     Function unknown  P75616  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load the annotation E Coli \n",
    "annotation = pd.read_csv(\"../data/Annotation/gene_info.csv\")\n",
    "annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the set of all known genes (E)\n",
    "all_genes = annotation.locus_tag.values \n",
    "\n",
    "# Define a set of all known regulators\n",
    "known_regulators = known_trn['RegulatorGeneName'].unique()\n",
    "\n",
    "# Map gene names to locus tags for comparable output with GENIE3 ouput\n",
    "annotation_map = annotation[annotation.gene_name.notna()].set_index(\"gene_name\")[\"locus_tag\"].to_dict()\n",
    "known_trn['targets_tags'] = known_trn[\"regulatedName\"].map(annotation_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over regulators and gene targets to load a graph from edges\n",
    "known_ecoli_trn = nx.Graph() \n",
    "for i in known_trn[['RegulatorGeneName', 'targets_tags']].itertuples():\n",
    "    known_ecoli_trn.add_edges_from([tuple([i[1], i[2]])]) \n",
    "    \n",
    "edges_ecoli_trn = known_ecoli_trn.edges\n",
    "\n",
    "# Find the neighbors of the regulators to define gene targets \n",
    "known_modules = []\n",
    "for reg in known_regulators:\n",
    "    known_modules.append(list(known_ecoli_trn.neighbors(reg)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulatoryGene</th>\n",
       "      <th>targetGene</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0573</td>\n",
       "      <td>b0574</td>\n",
       "      <td>0.041188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0575</td>\n",
       "      <td>b0574</td>\n",
       "      <td>0.039607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0574</td>\n",
       "      <td>b0575</td>\n",
       "      <td>0.039262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b2000</td>\n",
       "      <td>b2001</td>\n",
       "      <td>0.038284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>b1973</td>\n",
       "      <td>b0296</td>\n",
       "      <td>0.035970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulatoryGene targetGene    weight\n",
       "1          b0573      b0574  0.041188\n",
       "2          b0575      b0574  0.039607\n",
       "3          b0574      b0575  0.039262\n",
       "4          b2000      b2001  0.038284\n",
       "5          b1973      b0296  0.035970"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load the results from GENIE3 undirected run w/o regulator input\n",
    "genie3_undirected = pd.read_csv(\"../data/predicted_results/GENIE3_5_ET_Ksqrt_nTrees1000_nRegs0_2024-06-11.csv\", index_col=0)\n",
    "genie3_undirected.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulatoryGene</th>\n",
       "      <th>targetGene</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b1921</td>\n",
       "      <td>b1071</td>\n",
       "      <td>0.344796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b1921</td>\n",
       "      <td>b1924</td>\n",
       "      <td>0.340820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0571</td>\n",
       "      <td>b0572</td>\n",
       "      <td>0.339699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b1921</td>\n",
       "      <td>b1925</td>\n",
       "      <td>0.336633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>b1921</td>\n",
       "      <td>b1926</td>\n",
       "      <td>0.330987</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulatoryGene targetGene    weight\n",
       "1          b1921      b1071  0.344796\n",
       "2          b1921      b1924  0.340820\n",
       "3          b0571      b0572  0.339699\n",
       "4          b1921      b1925  0.336633\n",
       "5          b1921      b1926  0.330987"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load the results from GENIE3 directed run w/o regulator input\n",
    "genie3_directed = pd.read_csv(\"../data/predicted_results/GENIE3_4_RF_Ksqrt_nTrees1000_nRegs106_2024-06-26.csv\", index_col=0)\n",
    "genie3_directed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "directed_list = []\n",
    "undirected_list = []\n",
    "\n",
    "for i in range(100, 10000, 100):\n",
    "    # Load the predicted undirected network\n",
    "    ecoli_trn_undirected = nx.Graph()\n",
    "    ecoli_trn_undirected.add_edges_from([tuple(val) for val in genie3_undirected[['regulatoryGene', 'targetGene']].head(i).values])\n",
    "        # Load the predicted directed network\n",
    "    ecoli_trn_directed = nx.Graph()\n",
    "    ecoli_trn_directed.add_edges_from([tuple(val) for val in genie3_directed[['regulatoryGene', 'targetGene']].head(i).values])\n",
    "\n",
    "    # Filter the network for connections with known regulators\n",
    "    # Append neighborhood connected genes to the list of genesets\n",
    "    regulator_ids = annotation[annotation['gene_name'].isin(known_regulators)]['locus_tag'].values\n",
    "\n",
    "    modules_undirected = []\n",
    "    for reg in regulator_ids:\n",
    "        if reg in ecoli_trn_undirected.nodes:\n",
    "            module = list(ecoli_trn_undirected.neighbors(reg))\n",
    "            modules_undirected.append(module)\n",
    "\n",
    "    # Get the sets \n",
    "    undirected_modules_as_sets = [set(module) for module in modules_undirected]\n",
    "    \n",
    "    modules_directed = []\n",
    "    for reg in regulator_ids:\n",
    "        if reg in ecoli_trn_directed.nodes:\n",
    "            module = list(ecoli_trn_directed.neighbors(reg))\n",
    "            modules_directed.append(module)\n",
    "\n",
    "    # Get the sets \n",
    "    directed_modules_as_sets = [set(module) for module in modules_directed]\n",
    "\n",
    "    ModulesB = Modules(known_modules)\n",
    "    ModulesA_undirected = Modules(undirected_modules_as_sets)\n",
    "    ModulesA_directed = Modules(directed_modules_as_sets)\n",
    "\n",
    "    undirect_comp = ModulesComparison(ModulesA_undirected, ModulesB, all_genes).score(None)\n",
    "    direct_comp = ModulesComparison(ModulesA_directed, ModulesB, all_genes).score(None)\n",
    "\n",
    "    directed_list.append((i, direct_comp['precision'], direct_comp['recall'], direct_comp['recovery'], direct_comp['relevance'], direct_comp['F1rprr']))\n",
    "    undirected_list.append((i, undirect_comp['precision'], undirect_comp['recall'], undirect_comp['recovery'], undirect_comp['relevance'], undirect_comp['F1rprr']))\n",
    "\n",
    "directed_scores = pd.DataFrame(directed_list, columns=['edges', 'precision', 'recall', 'recovery', 'relevance', 'f1rprr'])\n",
    "undirected_scores = pd.DataFrame(undirected_list, columns=['edges', 'precision', 'recall', 'recovery', 'relevance', 'f1rprr'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edges</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>recovery</th>\n",
       "      <th>relevance</th>\n",
       "      <th>f1rprr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100</td>\n",
       "      <td>0.002694</td>\n",
       "      <td>0.004932</td>\n",
       "      <td>0.289750</td>\n",
       "      <td>0.056400</td>\n",
       "      <td>0.006721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>200</td>\n",
       "      <td>0.004052</td>\n",
       "      <td>0.009425</td>\n",
       "      <td>0.208927</td>\n",
       "      <td>0.102920</td>\n",
       "      <td>0.010888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>300</td>\n",
       "      <td>0.004430</td>\n",
       "      <td>0.010680</td>\n",
       "      <td>0.196678</td>\n",
       "      <td>0.116267</td>\n",
       "      <td>0.012011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>400</td>\n",
       "      <td>0.004725</td>\n",
       "      <td>0.011865</td>\n",
       "      <td>0.194786</td>\n",
       "      <td>0.125187</td>\n",
       "      <td>0.012943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>500</td>\n",
       "      <td>0.005154</td>\n",
       "      <td>0.013351</td>\n",
       "      <td>0.180458</td>\n",
       "      <td>0.132976</td>\n",
       "      <td>0.014186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>9500</td>\n",
       "      <td>0.011102</td>\n",
       "      <td>0.006452</td>\n",
       "      <td>0.095347</td>\n",
       "      <td>0.085162</td>\n",
       "      <td>0.014965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>9600</td>\n",
       "      <td>0.011178</td>\n",
       "      <td>0.006389</td>\n",
       "      <td>0.095063</td>\n",
       "      <td>0.085358</td>\n",
       "      <td>0.014913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>9700</td>\n",
       "      <td>0.011275</td>\n",
       "      <td>0.006360</td>\n",
       "      <td>0.095342</td>\n",
       "      <td>0.086113</td>\n",
       "      <td>0.014925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>9800</td>\n",
       "      <td>0.011290</td>\n",
       "      <td>0.006306</td>\n",
       "      <td>0.094598</td>\n",
       "      <td>0.085265</td>\n",
       "      <td>0.014845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>9900</td>\n",
       "      <td>0.011362</td>\n",
       "      <td>0.006264</td>\n",
       "      <td>0.094394</td>\n",
       "      <td>0.084712</td>\n",
       "      <td>0.014812</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    edges  precision    recall  recovery  relevance    f1rprr\n",
       "0     100   0.002694  0.004932  0.289750   0.056400  0.006721\n",
       "1     200   0.004052  0.009425  0.208927   0.102920  0.010888\n",
       "2     300   0.004430  0.010680  0.196678   0.116267  0.012011\n",
       "3     400   0.004725  0.011865  0.194786   0.125187  0.012943\n",
       "4     500   0.005154  0.013351  0.180458   0.132976  0.014186\n",
       "..    ...        ...       ...       ...        ...       ...\n",
       "94   9500   0.011102  0.006452  0.095347   0.085162  0.014965\n",
       "95   9600   0.011178  0.006389  0.095063   0.085358  0.014913\n",
       "96   9700   0.011275  0.006360  0.095342   0.086113  0.014925\n",
       "97   9800   0.011290  0.006306  0.094598   0.085265  0.014845\n",
       "98   9900   0.011362  0.006264  0.094394   0.084712  0.014812\n",
       "\n",
       "[99 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "directed_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edges</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>recovery</th>\n",
       "      <th>relevance</th>\n",
       "      <th>f1rprr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.208333</td>\n",
       "      <td>0.005518</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>200</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.012403</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>300</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.001191</td>\n",
       "      <td>0.438025</td>\n",
       "      <td>0.016057</td>\n",
       "      <td>0.000080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>400</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.001481</td>\n",
       "      <td>0.392011</td>\n",
       "      <td>0.018510</td>\n",
       "      <td>0.000570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>500</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.001481</td>\n",
       "      <td>0.392011</td>\n",
       "      <td>0.018510</td>\n",
       "      <td>0.000570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>9500</td>\n",
       "      <td>0.004315</td>\n",
       "      <td>0.010686</td>\n",
       "      <td>0.173925</td>\n",
       "      <td>0.113662</td>\n",
       "      <td>0.011768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>9600</td>\n",
       "      <td>0.004312</td>\n",
       "      <td>0.010732</td>\n",
       "      <td>0.169772</td>\n",
       "      <td>0.114237</td>\n",
       "      <td>0.011773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>9700</td>\n",
       "      <td>0.004455</td>\n",
       "      <td>0.010836</td>\n",
       "      <td>0.170448</td>\n",
       "      <td>0.114385</td>\n",
       "      <td>0.012071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>9800</td>\n",
       "      <td>0.004455</td>\n",
       "      <td>0.010836</td>\n",
       "      <td>0.170448</td>\n",
       "      <td>0.114385</td>\n",
       "      <td>0.012071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>9900</td>\n",
       "      <td>0.004453</td>\n",
       "      <td>0.010987</td>\n",
       "      <td>0.172779</td>\n",
       "      <td>0.115511</td>\n",
       "      <td>0.012120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    edges  precision    recall  recovery  relevance    f1rprr\n",
       "0     100   0.000001  0.000153  0.208333   0.005518  0.000005\n",
       "1     200   0.000008  0.000804  0.500000   0.012403  0.000033\n",
       "2     300   0.000020  0.001191  0.438025   0.016057  0.000080\n",
       "3     400   0.000159  0.001481  0.392011   0.018510  0.000570\n",
       "4     500   0.000159  0.001481  0.392011   0.018510  0.000570\n",
       "..    ...        ...       ...       ...        ...       ...\n",
       "94   9500   0.004315  0.010686  0.173925   0.113662  0.011768\n",
       "95   9600   0.004312  0.010732  0.169772   0.114237  0.011773\n",
       "96   9700   0.004455  0.010836  0.170448   0.114385  0.012071\n",
       "97   9800   0.004455  0.010836  0.170448   0.114385  0.012071\n",
       "98   9900   0.004453  0.010987  0.172779   0.115511  0.012120\n",
       "\n",
       "[99 rows x 6 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "undirected_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge count with highest combined F1rprr score:\n",
      "edges                   7900.000000\n",
      "precision_directed         0.010275\n",
      "recall_directed            0.007442\n",
      "recovery_directed          0.100343\n",
      "relevance_directed         0.093375\n",
      "f1rprr_directed            0.015850\n",
      "precision_undirected       0.004075\n",
      "recall_undirected          0.010815\n",
      "recovery_undirected        0.216044\n",
      "relevance_undirected       0.108215\n",
      "f1rprr_undirected          0.011373\n",
      "combined_f1rprr            0.027223\n",
      "Name: 78, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "#merge the data frames, compute combined scores, and find which number of edges has the highest combined score\n",
    "combined_scores = pd.merge(directed_scores, undirected_scores, on='edges', suffixes=('_directed', '_undirected'))\n",
    "combined_scores['combined_f1rprr'] = combined_scores['f1rprr_directed'] + combined_scores['f1rprr_undirected']\n",
    "max_combined_score = combined_scores.loc[combined_scores['combined_f1rprr'].idxmax()]\n",
    "\n",
    "print(\"Edge count with highest combined F1rprr score:\")\n",
    "print(max_combined_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the F1rprr scores, it would appear that 7900 edges is the optimal number."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "project",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
